xlators/mgmt/glusterd/src/glusterd-op-sm.c

```diff
@@ -3113,6 +3123,12 @@ glusterd_op_set_volume(dict_t *dict, char **errstr)
     gf_msg_debug(this->name, 0, "returning %d", ret);
     if (quorum_action)
         glusterd_do_quorum_action();
+    if (ret < 0 && count > 1) {
+        if (dict_reset(volinfo->dict) == 0)
+            dict_copy(volinfo_dict_orig, volinfo->dict);
```

Comment:
Why not this:

```suggestion
        dict_unref(volinfo->dict);
        volinfo->dict = volinfo_dict_orig;
        volinfo_dict_orig = NULL;
```

Reply:
Can we assume dict_unref will cleanup all dictionary members always? It will reset members only if refcount has become 0 in case of dict_reset we are 100% sure members will be reset.

Reply:
I am not too sure either, if by the time we reach glusterd_op_set_volume(), whether volinfo->dict has/ will have exactly only 1 reference. @xhernandez  what do you think?

Reply:
> Can we assume dict_unref will cleanup all dictionary members always? It will reset members only if refcount has become 0 in case of dict_reset we are 100% sure members will be reset.

If a dict has an additional reference it's because it's in use, so not destroying it seems the correct thing to do.

The problem here is what @itisravi says. If there are additional references means that someone is already using it, so replacing the dict is not a good option, we need to change its contents only.

I'm not sure if this can happen or not, but for the time being assuming that it could happen is the safest option. `dict_copy()` can still fail though, so we may not completely recover the original state in all cases, but the chances are very small.